sw/device/lib/testing/test_framework/freertos_port.S

```diff
@@ -15,6 +16,36 @@
 
 // -----------------------------------------------------------------------------
 
+/**
+ * Save MEPC to the stack.
+ *
+ * NOTE: this IRQ is synchronous, therefore, we must update the ISR return
+ * address to point to the instruction after the one that triggered this IRQ.
+ * Since we support the RISC-V compressed instructions extension, we need to
+ * check if the two least significant bits of the instruction are
+ * b11 (0x3), which means that the trapped instruction is not compressed,
+ * i.e., the trapped instruction is 32bits = 4bytes. Otherwise, the trapped
+ * instruction is 16bits = 2bytes.
+ */
+
+.balign 4
+.type process_mepc_on_synchronous_irq, @function
+process_mepc_on_synchronous_irq:
+  csrr t0, mepc
+  li t1, 0x3
+  and t2, t0, t1
+  beq t2, t1, L_32bit_trap_instr
+  addi t0, t0, PORT_HALF_WORD_SIZE
+  ret
+L_32bit_trap_instr:
```

Comment:
I might be missing something, but I don't see where you actually save the new value into the mepc CSR.

Reply:
Same for the branch above.

Reply:
This function only extracts the PC that follows it; the ISR is responsible for setting up the jump back to user code.

Reply:
Missed that it is done in `freertosIrqExit`, fair point.

Reply:
The mepc (or mepc + 1 instr) is saved to a FreeRTOS task's stack in the ISR entry subroutine (i.e., handler_exception, handler_irq_software, handler_irq_timer, handler_irq_external). Then, when the ISR exits (in freertosIrqExit), the mepc is written (using the value that was saved on the stack at ISR entry. The reason for this architecture is that a context switch could happen in an ISR (e.g., timer ISR for preemptive mode, and ecall exception handler for non-preemptive mode when "yield" is called). A context switch switches out the current task's stack pointer for the new task's stack pointer. So we have to wait for the exit subroutine so the correct stack is in place with the correct mepc to swap in. Does that make sense>